picker/leastloaded.go

```diff
@@ -138,24 +135,46 @@ func (h *leastLoadedConnHeap) update(allConns conn.Conns) {
 	for i, l := 0, allConns.Len(); i < l; i++ {
 		newMap[allConns.Get(i)] = struct{}{}
 	}
-
-	for i, l := 0, len(*h); i < l; i++ {
-		item := (*h)[i]
+	j := 0
+	slice := *h
+	// Remove items from slice that aren't in the new set of conns,
+	// compacting the slice as we go.
+	for i, item := range slice {
 		if _, ok := newMap[item.conn]; ok {
 			delete(newMap, item.conn)
+			if i != j {
+				item.index = j
+				(*h)[j] = item
+			}
+			j++
 		} else {
-			heap.Remove(h, item.index)
+			// If there are pending ops with this one, make sure it
+			// knows it's been evicted.
+			item.index = -1
 		}
 	}
-
-	for conn := range newMap {
-		heap.Push(h, &leastLoadedConnItem{conn: conn})
+	newLen := j + len(newMap)
+	if len(slice) > newLen {
+		// Make sure we don't leak memory with dangling pointers
+		// in unused regions of the slice.
+		for i := range slice[newLen:] {
+			slice[newLen+i] = nil
+		}
+	}
+	// Now add remaining new connections.
+	slice = slice[:j]
+	for cn := range newMap {
+		slice = append(slice, &leastLoadedConnItem{conn: cn, index: len(slice)})
 	}
+	*h = slice
+	// Re-heapify
+	heap.Init(h)
 }
 
-func (h *leastLoadedConnHeap) acquire() *leastLoadedConnItem {
+func (h *leastLoadedConnHeap) acquire(nextTieBreak uint64) *leastLoadedConnItem {
 	entry := (*h)[0]
 	entry.load++
+	entry.tieBreak = nextTieBreak
```

Comment:
I think we should be doing this here, instead of outside of thus function as was done before, so that the `Fix` call below takes this value into consideration.